This patch is the result of work initiated by Anton Denisov's 
<fire@kgpu.kamchatka.ru> feedback (and debug-patches).

There are three problems with apt-get source <pkgname>, this patch fixes two of them:

- rpmrecords.cc: SourcePackage() returns empty string because of incorrect 
  rpmlib's return code interpretation (fixed, now it return in most cases 
  the source package filename);
- srcrecords.cc: Find() called from apt-get's DoSource() doesn't match anything
  because it matches against source package name, but gets a filename as an 
  argument (in the case of the first branch in DoSource() when the return value 
  of SourcePackage() is used) (fixed: now matching is done for corresponding 
  fields).

This makes the first branch in DoSource() work which is almost all one may want.

9 July 2002
imz@altlinux.ru

A bad side of this patch may be that in the first branch (matching against
src.rpm name), it takes the full filenmame (including version-release) into
account while the second branch does only matching against the name (version
insensitive).

Ported to apt-0.5.4cnc9. (There is a separate apt-0.5.4cnc9-alt-getsrc-debug.patch
which helps to determine which of the branches is really used.)

imz@altlinux.ru, 28 Sep 2002.

--- apt-0.5.4cnc9/apt-pkg/deb/debsrcrecords.h.getsrc	2002-07-25 22:07:18 +0400
+++ apt-0.5.4cnc9/apt-pkg/deb/debsrcrecords.h	2002-10-29 00:45:07 +0300
@@ -34,6 +34,9 @@
    virtual bool Step() {iOffset = Tags.Offset(); return Tags.Step(Sect);};
    virtual bool Jump(unsigned long Off) {iOffset = Off; return Tags.Jump(Sect,Off);};
 
+   // These refer to the archive file for the Version
+   virtual string FileName() const {return Sect.FindS("Filename");};
+
    virtual string Package() const {return Sect.FindS("Package");};
    virtual string Version() const {return Sect.FindS("Version");};
    virtual string Maintainer() const {return Sect.FindS("Maintainer");};
--- apt-0.5.4cnc9/apt-pkg/rpm/rpmsrcrecords.h.getsrc	2002-08-09 00:07:33 +0400
+++ apt-0.5.4cnc9/apt-pkg/rpm/rpmsrcrecords.h	2002-10-29 00:45:07 +0300
@@ -44,6 +44,8 @@
    virtual bool Step(); 
    virtual bool Jump(unsigned long Off);
 
+   virtual string FileName() const;
+
    virtual string Package() const;
    virtual string Version() const;
    virtual string Maintainer() const;
--- apt-0.5.4cnc9/apt-pkg/rpm/rpmrecords.cc.getsrc	2002-08-19 20:37:48 +0400
+++ apt-0.5.4cnc9/apt-pkg/rpm/rpmrecords.cc	2002-10-29 01:02:58 +0300
@@ -174,7 +174,12 @@
 /* */
 string rpmRecordParser::SourcePkg()
 {
-   return "";
+   char *str;
+   int_32 count, type;
+   assert(HeaderP != NULL);
+   int rc = headerGetEntry(HeaderP, RPMTAG_SOURCERPM,
+			   &type, (void**)&str, &count);
+   return string(rc?str:"");
 }
 									/*}}}*/
 
--- apt-0.5.4cnc9/apt-pkg/rpm/rpmsrcrecords.cc.getsrc	2002-08-09 00:07:33 +0400
+++ apt-0.5.4cnc9/apt-pkg/rpm/rpmsrcrecords.cc	2002-10-29 00:57:41 +0300
@@ -143,6 +143,20 @@
    return true;
 }
 
+// RecordParser::FileName - Return the archive filename on the site	/*{{{*/
+// ---------------------------------------------------------------------
+/* */
+string rpmSrcRecordParser::FileName() const
+{
+   char *str;
+   int_32 count, type;
+   assert(HeaderP != NULL);
+   int rc = headerGetEntry(HeaderP, CRPMTAG_FILENAME,
+			   &type, (void**)&str, &count);
+   return string(rc?str:"");
+}
+									/*}}}*/
+
 string rpmSrcRecordParser::Package() const
 {
    char *str;
--- apt-0.5.4cnc9/apt-pkg/srcrecords.cc.getsrc	2002-07-23 21:54:50 +0400
+++ apt-0.5.4cnc9/apt-pkg/srcrecords.cc	2002-10-29 00:45:07 +0300
@@ -82,7 +82,7 @@
 									/*}}}*/
 // SrcRecords::Find - Find the first source package with the given name	/*{{{*/
 // ---------------------------------------------------------------------
-/* This searches on both source package names and output binary names and
+/* This searches on both source package filenames and output binary names and
    returns the first found. A 'cursor' like system is used to allow this
    function to be called multiple times to get successive entries */
 pkgSrcRecords::Parser *pkgSrcRecords::Find(const char *Package,bool SrcOnly)
@@ -92,6 +92,8 @@
    
    while (true)
    {
+      // DEBUG:
+      //std::cerr << "start loop" << std::endl;
       // Step to the next record, possibly switching files
       while ((*Current)->Step() == false)
       {
@@ -107,7 +109,7 @@
 	 return 0;
 
       // Source name hit
-      if ((*Current)->Package() == Package)
+      if ((*Current)->FileName() == Package)
 	 return *Current;
       
       if (SrcOnly == true)
--- apt-0.5.4cnc9/apt-pkg/srcrecords.h.getsrc	2002-07-25 22:07:18 +0400
+++ apt-0.5.4cnc9/apt-pkg/srcrecords.h	2002-10-29 00:45:07 +0300
@@ -66,6 +66,8 @@
       virtual unsigned long Offset() = 0;
       virtual string AsStr() = 0;
       
+      virtual string FileName() const = 0;
+
       virtual string Package() const = 0;
       virtual string Version() const = 0;
       virtual string Maintainer() const = 0;
